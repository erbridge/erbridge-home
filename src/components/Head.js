import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Helmet } from 'react-helmet';
import { Route } from 'react-router-dom';

import routePropType from '../prop-types/routePropType';

class Head extends Component {
  static propTypes = {
    routes: PropTypes.arrayOf(routePropType).isRequired,
  };

  render() {
    const { routes } = this.props;

    return routes.map((route, index) => (
      <Route key={route.name || index} path={route.path} exact>
        <Helmet>
          <title>
            {route.title ? `${route.title} | ` : ''}
            erbridge
          </title>
        </Helmet>
      </Route>
    ));
  }
}

export default Head;
