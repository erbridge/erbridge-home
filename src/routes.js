import moment from 'moment';

import blogPosts from './blog';

import home from './pages/home.md';

const sanitizePathComponent = rawComponent =>
  encodeURI(
    rawComponent
      .toLowerCase()
      .replace(/\s/g, '-')
      .replace(/--+/g, '-')
      .replace(/:|#/g, ''),
  );

const getBlogRoutePath = (slug, title) =>
  `/blog/${sanitizePathComponent(slug || title)}`;

const formatDate = (date, format) => {
  if (!date) {
    return null;
  }

  date = moment(date);

  if (!date.isValid()) {
    return null;
  }

  return date.format(format);
};

export const homeRoute = {
  name: 'home',
  link: 'Home',
  path: '/',
  exact: true,
  navExact: true,
  title: home.title,
  styles: home.styles,
  content: home.__content,
};

const sortedBlogPosts = [...blogPosts];

sortedBlogPosts.sort((a, b) => moment(b.date) - moment(a.date));

const blogRoutes = sortedBlogPosts.map(
  ({ __content, categories, date, slug, styles, subtitle, title }) => ({
    path: getBlogRoutePath(slug, title),
    exact: true,
    title,
    subtitle,
    date: formatDate(date, 'DD MMMM YYYY'),
    categories,
    styles,
    content: __content,
  }),
);

const blogCategoryRoutes = [
  {
    name: 'blog__all',
    path: '/blog/all',
    exact: true,
    title: 'All Posts',
    routes: blogRoutes,
  },
  {
    name: 'blog__gamedev',
    path: '/blog/gamedev',
    exact: true,
    title: 'Game Development',
    routes: blogRoutes.filter(
      ({ categories }) => categories && categories.indexOf('gamedev') !== -1,
    ),
  },
  {
    name: 'blog__journals',
    path: '/blog/journals',
    exact: true,
    title: 'Roleplaying Journals',
    routes: blogRoutes.filter(
      ({ categories }) =>
        categories &&
        categories.indexOf('roleplaying') !== -1 &&
        categories.indexOf('journal') !== -1,
    ),
  },
];

const blogRoute = {
  name: 'blog',
  link: 'Blog',
  path: '/blog',
  exact: true,
  title: 'Blog',
  routes: blogCategoryRoutes,
};

export const redirectedRoutes = [
  {
    path: '/about',
    to: homeRoute.path,
    exact: true,
  },
  {
    path: '/blog/skald',
    to: blogRoute.path,
    exact: true,
  },
  ...sortedBlogPosts
    .map(post => ({ ...post, date: formatDate(post.date, 'YYYY/MM/DD') }))
    .filter(({ date }) => date)
    .map(({ date, oldSlug, slug, title }) => ({
      path: `/blog/${date}/${sanitizePathComponent(oldSlug || slug || title)}`,
      to: getBlogRoutePath(slug, title),
      exact: true,
    })),
];

export const topRoutes = [homeRoute, blogRoute];

export default [...topRoutes, ...blogCategoryRoutes, ...blogRoutes];
