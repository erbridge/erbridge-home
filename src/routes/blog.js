import blogPosts from '../pages/blog';

import {
  formatDate,
  generateRoutes,
  getRoutePath,
  sanitizePathComponent,
} from './helpers';

export const {
  categoryRoutes,
  redirectedRoutes,
  route,
  routes,
} = generateRoutes('blog', 'blog', 'Blog', 'Blog', blogPosts, {
  categories: [
    {
      name: 'all',
      path: 'all',
      link: 'All',
      title: 'All Posts',
      routeFilter: route => route,
    },
    {
      name: 'gamedev',
      path: 'gamedev',
      link: 'Game Development',
      title: 'Game Development Posts',
      routeFilter: ({ categories }) =>
        categories && categories.indexOf('gamedev') !== -1,
    },
    {
      name: 'journals',
      path: 'journals',
      link: 'Roleplaying Journals',
      title: 'Roleplaying Journals',
      routeFilter: ({ categories }) =>
        categories &&
        categories.indexOf('roleplaying') !== -1 &&
        categories.indexOf('journal') !== -1,
    },
    {
      name: 'uncategorized',
      path: 'uncategorized',
      link: 'Uncategorized',
      title: 'Uncategorized Posts',
      routeFilter: ({ categories }) =>
        !categories ||
        (categories.indexOf('gamedev') === -1 &&
          categories.indexOf('roleplaying') === -1 &&
          categories.indexOf('journal') === -1),
    },
  ],
  extraRedirects: [
    {
      path: '/blog/skald',
      to: getRoutePath('blog'),
    },
  ],
  generateRedirects: posts =>
    posts
      .map(post => ({ ...post, date: formatDate(post.date, 'YYYY/MM/DD') }))
      .filter(({ date }) => date)
      .map(({ date, oldSlug, slug, title }) => ({
        path: `/blog/${date}/${sanitizePathComponent(
          oldSlug || slug || title,
        )}`,
        to: getRoutePath('blog', slug, title),
      })),
});
