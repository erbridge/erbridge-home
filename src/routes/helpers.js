import moment from 'moment';

export const formatDate = (date, format) => {
  if (!date) {
    return null;
  }

  date = moment(date);

  if (!date.isValid()) {
    return null;
  }

  return date.format(format);
};

export const sanitizePathComponent = rawComponent =>
  encodeURI(
    rawComponent
      .toLowerCase()
      .replace(/\s/g, '-')
      .replace(/--+/g, '-')
      .replace(/:|#|\(|\)|@/g, ''),
  );

export const getRoutePath = (prefix, slug, title) =>
  `/${prefix}/${sanitizePathComponent(slug || title)}`;
