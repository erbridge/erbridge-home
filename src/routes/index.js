import home from '../pages/home.md';

import {
  categoryRoutes as blogCategoryRoutes,
  route as blogRoute,
  routes as blogRoutes,
  redirectedRoutes as redirectedBlogRoutes,
} from './blog';
import {
  route as fictionRoute,
  routes as fictionRoutes,
  redirectedRoutes as redirectedFictionRoutes,
} from './fiction';

export const homeRoute = {
  name: 'home',
  link: 'Home',
  path: '/',
  exact: true,
  navExact: true,
  title: home.title,
  styles: home.styles,
  content: home.__content,
};

export const redirectedRoutes = [
  {
    path: '/about',
    to: homeRoute.path,
    exact: true,
  },
  ...redirectedBlogRoutes,
  ...redirectedFictionRoutes,
];

export const topRoutes = [homeRoute, fictionRoute, blogRoute];

export default [
  ...topRoutes,
  ...blogCategoryRoutes,
  ...blogRoutes,
  ...fictionRoutes,
];
