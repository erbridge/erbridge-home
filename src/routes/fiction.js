import fictionPosts from '../pages/fiction';

import { generateRoutes } from './helpers';

export const {
  categoryRoutes,
  redirectedRoutes,
  route,
  routes,
} = generateRoutes('fiction', 'fiction', 'Fiction', 'Fiction', fictionPosts, {
  categories: [
    {
      name: 'all',
      path: 'all',
      link: 'All',
      title: 'All Fiction',
      routeFilter: Boolean,
    },
    {
      name: 'poetry',
      path: 'poetry',
      title: 'Poetry',
      routeFilter: ({ categories }) =>
        categories && categories.indexOf('poetry') !== -1,
    },
    {
      name: 'screenplays',
      path: 'screenplays',
      title: 'Screenplays',
      routeFilter: ({ categories }) =>
        categories && categories.indexOf('screenplay') !== -1,
    },
    {
      name: 'short-stories',
      path: 'short-stories',
      title: 'Short Stories',
      routeFilter: ({ categories }) =>
        categories && categories.indexOf('short-story') !== -1,
    },
    {
      name: 'generative',
      path: 'generative',
      link: 'Generative',
      title: 'Generative Fiction',
      routeFilter: ({ categories }) =>
        categories && categories.indexOf('generative') !== -1,
    },
    {
      name: 'uncategorized',
      path: 'uncategorized',
      link: 'Uncategorized',
      title: 'Uncategorized Fiction',
      routeFilter: ({ categories }) =>
        !categories ||
        (categories.indexOf('poetry') === -1 &&
          categories.indexOf('screenplay') === -1 &&
          categories.indexOf('short-story') === -1 &&
          categories.indexOf('generative') === -1),
    },
  ],
});
