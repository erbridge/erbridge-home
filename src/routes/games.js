import gamesPosts from '../pages/games';

import { generateRoutes, getRoutePath } from './helpers';

export const { redirectedRoutes, route, routes } = generateRoutes(
  'games',
  'games',
  'Games',
  'Games',
  gamesPosts,
  {
    extraRedirects: [
      {
        path: '/fatal-attraction',
        to: getRoutePath('games', 'fatal-attraction'),
      },
    ],
  },
);
