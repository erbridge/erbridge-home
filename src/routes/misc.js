import miscPosts from '../pages/misc';

import { generateRoutes, getRoutePath } from './helpers';

export const {
  categoryRoutes,
  redirectedRoutes,
  route,
  routes,
} = generateRoutes('misc', 'misc', 'Misc', 'Misc', miscPosts, {
  categories: [
    {
      name: 'all',
      path: 'all',
      link: 'All',
      title: 'All Misc Projects',
      routeFilter: route => route,
    },
    {
      name: 'bots',
      path: 'bots',
      title: 'Bots',
      routeFilter: ({ categories }) =>
        categories && categories.indexOf('bot') !== -1,
    },
    {
      name: 'uncategorized',
      path: 'uncategorized',
      link: 'Uncategorized',
      title: 'Uncategorized Misc Projects',
      routeFilter: ({ categories }) =>
        !categories || categories.indexOf('bot') === -1,
    },
  ],
  extraRedirects: [
    {
      path: '/tools/gotwit',
      to: getRoutePath('misc', 'gotwit'),
    },
    {
      path: '/tools/iris',
      to: getRoutePath('misc', 'iris'),
    },
    {
      path: '/tools/itchy-electron',
      to: getRoutePath('misc', 'itchy-electron'),
    },
    {
      path: '/tools/lime-text',
      to: getRoutePath('misc', 'limetext'),
    },
    {
      path: '/tools/nqr',
      to: getRoutePath('misc', 'nqr'),
    },
  ],
});
