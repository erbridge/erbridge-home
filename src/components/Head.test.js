import { shallow } from 'enzyme';
import React from 'react';

import routes from '../routes';

import Head from './Head';

jest.mock('../pages/blog');
jest.mock('../pages/fiction');
jest.mock('../pages/games');
jest.mock('../pages/home.mdx', () => ({ meta: {}, default: () => 'home.mdx' }));
jest.mock('../pages/misc');
jest.mock('../pages/web');

it('renders correctly', () => {
  expect(shallow(<Head routes={routes} />)).toMatchSnapshot();
});
