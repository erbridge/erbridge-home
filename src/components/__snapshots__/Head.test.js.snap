// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <Route
    exact={true}
    key="home"
    path="/"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        erbridge
      </title>
    </HelmetWrapper>
  </Route>,
  <Route
    exact={true}
    key="games"
    path="/games"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        Games | 
        erbridge
      </title>
    </HelmetWrapper>
  </Route>,
  <Route
    exact={true}
    key="web"
    path="/web"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        Web | 
        erbridge
      </title>
    </HelmetWrapper>
  </Route>,
  <Route
    exact={true}
    key="misc"
    path="/misc"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        Misc | 
        erbridge
      </title>
    </HelmetWrapper>
  </Route>,
  <Route
    exact={true}
    key="fiction"
    path="/fiction"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        Fiction | 
        erbridge
      </title>
    </HelmetWrapper>
  </Route>,
  <Route
    exact={true}
    key="blog"
    path="/blog"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        Blog | 
        erbridge
      </title>
    </HelmetWrapper>
  </Route>,
  <Route
    exact={true}
    key="books"
    path="/books"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        erbridge
      </title>
    </HelmetWrapper>
  </Route>,
]
`;
